File: sach/de3303.py
//DE3303.PY - Tran Huu Nam - hsgtin.vn - 2024
def convert_char(s):
    cs = "0123456789ABCDEF"
    x = 0
    for i in range(4):
        if s[i] == '1':
            x += 2**(3-i)
    return cs[x]
def bin2hex(s):
    while len(s) % 4 != 0:
        s = '0' + s
    p = ''
    for i in range(0,len(s),4):
        p += convert_char(s[i:i+4])
    while p[0] == '0':
        p = p[1:]
    return p
n = int(input())
a=[]
for x in range(n):
    s = input().split()
    a.append(s)
dp = [ ['']*(n+1) for j in range(n+1) ]
for i in range(1,n+1):
    dp[i][1] = dp[i-1][1] + a[i-1][0]
    dp[1][i] = dp[1][i-1] + a[0][i-1]
for i in range(2,n+1):
    for j in range(2,n+1):
        dp[i][j] = max(dp[i-1][j], dp[i][j-1]) + a[i-1][j-1]
print(bin2hex(dp[n][n]))

File: sach/de3303.cpp
//DE3303.CPP - Tran Huu Nam - hsgtin.vn - 2024
#include <bits/stdc++.h>
#define ll long long
using namespace std;
const ll N=101;
string cs="0123456789ABCDEF";
ll n;
char a[N][N];   string F[N][N];
void bin2hex(string s){
    while (s[0]=='0') s.erase(0,1);
    while (s.length()%4!=0) s="0"+s;
    for (int i=0; i<s.length(); i+=4) {
        int p=0;
        for(int j=i; j<=i+3; j++) p=p*2+s[j]-48;
        cout<<cs[p];
    }
}
int main () {
    ios_base::sync_with_stdio(false); cin.tie(NULL);
    cin >> n;
    for (int i=1;i<=n;i++)
      for (int j=1;j<=n;j++)
        cin >> a[i][j];
    for (int i=0;i<=n;i++) F[0][i]=F[i][0]="";
    for (int i=1;i<=n;i++)
      for (int j=1;j<=n;j++)
        F[i][j]=max(F[i-1][j],F[i][j-1])+a[i][j];
    bin2hex(F[n][n]);
 }

